Set ddns-hostname instead of option host-name for host declarations

treydock opened the pull request:
This is a non-backwards compatible change, at least in terms of the change that would occur to a DHCP server's config files.  This change is partially to incorporate what's done by theforeman-dhcp into this module (theforeman/puppet-dhcp#21).

Per the man page for dhcpd.conf it seems like ddns-hostname is used first when determing a client's hostname.

       The DHCP server determines the client’s hostname by first looking for a ddns-hostname configuration option, and using that if it is present.  If no such option is present, the server looks for a valid hostname in the FQDN option sent by
       the client.  If one is found, it is used; otherwise, if the client sent a host-name option, that is used.  Otherwise, if there is a host declaration that applies to the client, the name from that declaration will be used.   If  none  of
       these applies, the server will not have a hostname for the client, and will not be able to do a DNS update.

And

            It should be noted here that most DHCP clients completely ignore the host-name option sent by the DHCP server, and there is no way to configure them not to do this.   So you generally have a choice of either not having any hostname
            to client IP address mapping that the client will recognize, or doing DNS updates.   It is beyond the scope of this document to describe how to make this determination.

Be happy to refactor this PR to simply add `ddns-hostname` while also still setting `option host-name`.

I do currently have working DDNS with `ddns-hostname` being set.  If someone can confirm that DDNS still works when setting `option host-name` then this PR may be of no real value.

jonnytpuppet commented:
@treydock thanks for your contribution!

I think in order to see your PR merged more quickly, it would be better if you just created a new property for ddns-hostname in the module, in order to maintain backwards-compatibility.

Otherwise, we would run the risk of having to make a major release based on this, and potentially breaking other configurations.

Thoughts?

Thanks!

jonnytpuppet commented:
Also, please remember to update the README and squash your commits.

Thanks!

hunner commented:
@treydock Could you rebase this so it can be merged? Thanks!

cmurphy commented:
I think it's fine to be backwards incompatible since this is a 0.x module. With #60 merged users can still set `option host-name` if they need to.

treydock commented:
@hunner I've rebased and also refactored commit to make `ddns-hostname` an optional value set via boolean parameter.

Let me know if I should refactor again to make `ddns-hostname` replace `option host-name`.

hunner commented:
@treydock I actually like having the title configure `ddns-hostname` because we merged #60 and that can be used to set `options host-name` so could you modify it to just do that? No boolean needed.

Also, it looks like the changes from #58 came into this PR during the rebase. Could you fix that so that they are still separate PRs?

treydock commented:
@hunner I've restored the original behavior of this PR which is to replace the setting of `option host-name` with `ddns-hostname`.  Also removed changes from #58 which I think made their way in due to my IDE caching file contents when switching branches.

hunner commented:
@treydock Hmm. Looks like the readme changes are now missing...

treydock commented:
@hunner The README changes for this PR I had made were to document the previous `set_ddns_hostname` parameter.  Not sure what README changes would be appropriate for this PR since what's changed is how the module defines a DHCP host entry.
